Support for multiple domains in Access Control Allow Origin

Shiranuit opened the pull request:
<!--
  Thank you for submitting a Pull Request. Please:
    - Read our CONTRIBUTING guidelines.
      https://github.com/kuzzleio/kuzzle/blob/master/CONTRIBUTING.md
    - Associate an issue with the Pull Request.
    - IMPORTANT - Add the corresponding "changelog:xxx" label to your PR.
-->

<!--- This template is optional. -->

## What does this PR do ?

This PR adds the support of multiple values in the config `http.accessControlAllowOrigin`
By using coma separated value in a string like this:
```json
{
  "http": {
    "accessControlAllowOrigin": "foo, bar, localhost"
  }
}
```
Or by using an  array directly
```json
{
  "http": {
    "accessControlAllowOrigin": [
      "foo",
      "bar",
      "localhost"
    ]
  }
}
```

When a request containing the header `Origin` is received, the funnel will first check if the config `http.accessControlAllowOrigin` contains `*`,  or if the Origin value is contained in `http.accessControlAllowOrigin` in this case the request continue to follow the validation process (`checkRights`, etc...), otherwise the request will be rejected with an `unauthorized_origin` error.

Request not containing the header `Origin` are not affected by the modification described above and follow the usual validation process (`checkRights`, etc...).

xbill82 commented on lib/cluster/node.js:
```diff
@@ -376,6 +376,15 @@ class ClusterNode {
+        const duplicate = nodes.filter(node => node.ip === this.ip);
+
+        if (duplicate.length > 0) {
+          global.kuzzle.log.error(`[FATAL] Another node share the same IP address as this one (${this.ip}): ${duplicate[0].id}. Shutting down.`);
```
This looks like @scottinet 's hotfix... Wasn't this living on a separate PR?

Shiranuit replied:
IDK, I just based this branch on 2-dev, it seems pretty weird

xbill82 replied:
Ok, I thought you merged his branch while still in PR... BTW yes it's really weird git sees these lines as diff... Did you merge 2-dev on this branch lately? Even if it should make no difference, sinc diffs on GH are performed on the latest commit of the destination branch... OH! Wait! I see your destination branch is `master`!

Shiranuit replied:
Ahh, you're right, I did put the wrong branch, I will changed it to 2-dev 👍

scottinet replied:
Oh, right. It's fortunate you raised an issue about this one @xbill82: I saw this too, but just thought that it was some rebase mishap or something like that, and paid it no mind :sweat_smile:

Shiranuit commented:
I did some minor corrections:
- Updated docker-compose (added admin console to Access-Control-Allow-Origin)
- Funnel: Changed the way the unauthorized_origin error is returned because it was causing a malformed HTTP Response
- Funnel & AuthController -> replace `request.input.args.cookieAuth` with `request.getBoolean('cookieAuth')`

scottinet commented:
@Shiranuit > the CI fails because you need to update the errors documentation: run `node doc/build-error-codes.js` and then push the changes

Leodau commented on lib/config/index.js:
```diff
@@ -212,4 +214,22 @@ function checkClusterOptions (config) {
   assert(typeof cfg.ip === 'string', '[FATAL] kuzzlerc.cluster.ip: value must be a string');
 }
 
+function preprocessHttpOptions (config) {
+  const cfg = config.http;
+
+  if (cfg === undefined) {
+    return;
+  }
+
+  if (typeof cfg.accessControlAllowOrigin === 'string') {
+    cfg.accessControlAllowOrigin =
+      cfg.accessControlAllowOrigin
+        .split(',')
+        .map(value => value.trim());
+  }
+
+  config.internal.allowAllOrigins = cfg.accessControlAllowOrigin.includes('*'); // Stored to avoid doing includes multiple times later
+  config.internal.cookieAuthentication = !cfg.accessControlAllowOrigin.includes('*'); // If there is no wild card, cookieAuthentication is enabled
+}
```
```suggestion
function preprocessHttpOptions (config) {
  const httpConfig = config.http;

  if (httpConfig === undefined) {
    return;
  }

  if (typeof httpConfig.accessControlAllowOrigin === 'string') {
    httpConfig.accessControlAllowOrigin =
      httpConfig.accessControlAllowOrigin
        .split(',')
        .map(value => value.trim());
  }

  // Stored to avoid doing includes multiple times later
  config.internal.allowAllOrigins = httpConfig.accessControlAllowOrigin.includes('*');

  // If there is no wild card, cookieAuthentication is enabled
  config.internal.cookieAuthentication = !httpConfig.accessControlAllowOrigin.includes('*');
}
```

Leodau commented on lib/config/index.js:
```diff
@@ -50,13 +50,14 @@ function load () {
   checkWebSocketOptions(config);
   checkClusterOptions(config);
-
+  
```
```suggestion

```

Leodau commented on lib/api/funnel.js:
```diff
@@ -263,6 +264,21 @@ class Funnel {
       return 1;
     }
 
+    const httpCfg = global.kuzzle.config.http;
+    if (request.input.headers
+      && has(request.input.headers, 'origin')
+      && !global.kuzzle.config.internal.allowAllOrigins
+      && !httpCfg.accessControlAllowOrigin.includes(request.input.headers.origin) // If the domain is not present in the list
```
Optional Nit
```suggestion
    const httpConfig = global.kuzzle.config.http;
    if (request.input.headers
      && has(request.input.headers, 'origin')
      && !global.kuzzle.config.internal.allowAllOrigins
      && !httpConfig.accessControlAllowOrigin.includes(request.input.headers.origin) // If the domain is not present in the list
```

Leodau replied:
I saw that we are using the `httpCfg` style in other places (protocol), you can ignore this comment if you think httpCfg > httpConfig.

Leodau commented on lib/core/network/httpRouter/index.js:
```diff
@@ -156,6 +155,21 @@ class Router {
       routeHandler.request.response.setHeaders(this.defaultHeaders, true);
 
+      
+      if (message.headers && message.headers.origin) {
```
Seems like duplicated code in **L158** & **L186** -
Is the intention to have explicit duplicated code or could this be abstracted?
